packages/test/test-utils/src/testObjectProvider.ts

```diff
@@ -474,10 +591,339 @@ export class TestObjectProvider implements ITestObjectProvider {
+	private _logger: EventAndErrorTrackingLogger | undefined;
+	private readonly _documentIdStrategy: IDocumentIdStrategy;
+	private _documentServiceFactory: IDocumentServiceFactory | undefined;
+	private _urlResolver: IUrlResolver | undefined;
+	// Since documentId doesn't change we can only create/make one container. Call the load functions instead.
+	private _documentCreated = false;
+	// `_loadCount` is used to alternate which version we load the next container with.
+	// If loadCount is even then we will load with the "create" version, and if odd we load with the "load" version.
+	// After each test we will reset loadCount to 0 to ensure we always create the first container with the create version.
+	// Note: This will only affect tests that load a container more than two times.
+	private _loadCount: number = 0;
+
+	constructor(
+		private readonly LoaderConstructorForCreating: typeof Loader,
+		private readonly LoaderConstructorForLoading: typeof Loader,
+		private readonly driverForCreating: ITestDriver,
+		private readonly driverForLoading: ITestDriver,
+		private readonly createFluidEntryPointForCreating: (
+			testContainerConfig?: ITestContainerConfig,
+		) => fluidEntryPoint,
+		private readonly createFluidEntryPointForLoading: (
+			testContainerConfig?: ITestContainerConfig,
+		) => fluidEntryPoint,
+	) {
+		this._documentIdStrategy = getDocumentIdStrategy(driverForCreating.type);
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.logger}
+	 */
+	public get logger(): EventAndErrorTrackingLogger {
+		if (this._logger === undefined) {
+			this._logger = new EventAndErrorTrackingLogger(
+				createChildLogger({
+					logger: getTestLogger?.(),
+				}),
+			);
+		}
+		return this._logger;
+	}
+
+	private set logger(logger: EventAndErrorTrackingLogger) {
+		this._logger = logger;
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.documentServiceFactory}
+	 */
+	public get documentServiceFactory() {
+		if (!this._documentServiceFactory) {
+			this._documentServiceFactory = this.driverForCreating.createDocumentServiceFactory();
+		}
+		return this._documentServiceFactory;
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.urlResolver}
+	 */
+	public get urlResolver() {
+		if (!this._urlResolver) {
+			this._urlResolver = this.driverForCreating.createUrlResolver();
+		}
+		return this._urlResolver;
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.documentId}
+	 */
+	public get documentId() {
+		return this._documentIdStrategy.get();
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.defaultCodeDetails}
+	 */
+	public get defaultCodeDetails() {
+		return defaultCodeDetails;
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.opProcessingController}
+	 */
+	public get opProcessingController(): IOpProcessingController {
+		return this._loaderContainerTracker;
+	}
+
+	/**
+	 * {@inheritDoc ITestObjectProvider.driver}
+	 */
+	public get driver(): ITestDriver {
+		return this._loadCount % 2 === 0 ? this.driverForCreating : this.driverForLoading;
```

Comment:
I think `this._loadCount % 2 === 0` means "The next loader we create will use ____" here and throughout.  Consider a private method to put a name on this, e.g. `private nextLoaderShouldCreate(): boolean` so it's not repeating the mysterious math throughout.

That said, is it correct to say we're toggling between creating and loading?  Or is this more of `nextLoaderShouldBeOld` or `nextLoaderShouldBeTypeA`?  It doesn't seem like we're actually restricting those Loaders to create/load respectively.

Reply:
Yeah I think it is more accurate at this point to say `loaderForOld` and `loaderForNew` or something like that. I kept  using this naming format because I wasn't initially sure if we would end up toggling between versions like this. Might be worth swapping out all the names?